pkg/quarterdeck/keygen/keygen.go

```diff
@@ -0,0 +1,69 @@
+/*
+Package keygen provides functionality for generating API client IDs and secrets.
+*/
+package keygen
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"strings"
+)
+
+const (
+	alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	alphanum = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+	idxbits  = 6
+	idxmask  = 1<<idxbits - 1
+	idxmax   = 63 / idxbits
+)
+
+func Alpha(n int) string {
+	sb := strings.Builder{}
+	sb.Grow(n)
+
+	for i, cache, remain := n-1, CryptoRandInt(), idxmax; i >= 0; {
+		if remain == 0 {
+			cache, remain = CryptoRandInt(), idxmax
+		}
+
+		if idx := int(cache & idxmask); idx < len(alphabet) {
+			sb.WriteByte(alphabet[idx])
+			i--
+		}
+
+		cache >>= idxbits
+		remain--
+	}
+
+	return sb.String()
+}
+
+func AlphaNumeric(n int) string {
+	sb := strings.Builder{}
+	sb.Grow(n)
+
+	for i, cache, remain := n-1, CryptoRandInt(), idxmax; i >= 0; {
+		if remain == 0 {
+			cache, remain = CryptoRandInt(), idxmax
+		}
+
+		if idx := int(cache & idxmask); idx < len(alphanum) {
+			sb.WriteByte(alphanum[idx])
+			i--
+		}
+
+		cache >>= idxbits
+		remain--
+	}
```

Comment:
Any reason why the duplicated code between the two methods can't be pulled down into a helper?

Reply:
One of the key points of the efficiency of the method was the reduction of allocations; I was worried that a helper would cause a copy of the alphabet string into the secondary method - but let me benchmark it to see what happens.

Original benchmarks:

```
goos: darwin
goarch: arm64
pkg: github.com/rotationalio/ensign/pkg/quarterdeck/keygen
BenchmarkAlpha16-10            	 1481076	       812.0 ns/op	      34 B/op	       3 allocs/op
BenchmarkAlpha64-10            	  414356	      2870 ns/op	     130 B/op	       9 allocs/op
BenchmarkAlpha256-10           	  109056	     11036 ns/op	     511 B/op	      32 allocs/op
BenchmarkAlphaNumeric16-10     	 1755441	       688.8 ns/op	      32 B/op	       3 allocs/op
BenchmarkAlphaNumeric64-10     	  497796	      2374 ns/op	     120 B/op	       8 allocs/op
BenchmarkAlphaNumeric256-10    	  130353	      9130 ns/op	     471 B/op	      27 allocs/op
BenchmarkCryptoRandInt-10      	 3710748	       321.8 ns/op	       8 B/op	       1 allocs/op
BenchmarkRandInt-10            	518337447	         2.287 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/rotationalio/ensign/pkg/quarterdeck/keygen	12.423s
```

Also just for funsies; benchmarks with `rand.Int63` instead of `CryptoRandInt` (it's a BIG difference!):

```
goos: darwin
goarch: arm64
pkg: github.com/rotationalio/ensign/pkg/quarterdeck/keygen
BenchmarkAlpha16-10            	16973916	        70.01 ns/op	      16 B/op	       1 allocs/op
BenchmarkAlpha64-10            	 5147656	       233.8 ns/op	      64 B/op	       1 allocs/op
BenchmarkAlpha256-10           	 1376426	       871.4 ns/op	     256 B/op	       1 allocs/op
BenchmarkAlphaNumeric16-10     	28164853	        42.21 ns/op	      16 B/op	       1 allocs/op
BenchmarkAlphaNumeric64-10     	 8398689	       142.1 ns/op	      64 B/op	       1 allocs/op
BenchmarkAlphaNumeric256-10    	 2150409	       559.9 ns/op	     256 B/op	       1 allocs/op
BenchmarkCryptoRandInt-10      	 3735400	       322.8 ns/op	       8 B/op	       1 allocs/op
BenchmarkRandInt-10            	528870255	         2.264 ns/op	       0 B/op	       0 allocs/op
```

Reply:
Looks like it doesn't really matter, so I've used the helper function method:

```
goos: darwin
goarch: arm64
pkg: github.com/rotationalio/ensign/pkg/quarterdeck/keygen
BenchmarkAlpha16-10            	 1492770	       808.6 ns/op	      34 B/op	       3 allocs/op
BenchmarkAlpha64-10            	  416186	      2881 ns/op	     130 B/op	       9 allocs/op
BenchmarkAlpha256-10           	  108795	     10999 ns/op	     511 B/op	      32 allocs/op
BenchmarkAlphaNumeric16-10     	 1751092	       685.6 ns/op	      32 B/op	       3 allocs/op
BenchmarkAlphaNumeric64-10     	  499417	      2386 ns/op	     120 B/op	       8 allocs/op
BenchmarkAlphaNumeric256-10    	  130398	      9160 ns/op	     471 B/op	      27 allocs/op
BenchmarkCryptoRandInt-10      	 3628449	       323.0 ns/op	       8 B/op	       1 allocs/op
BenchmarkRandInt-10            	527885359	         2.242 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/rotationalio/ensign/pkg/quarterdeck/keygen	12.407s
```